src/app/components/shared/date-time-filter/filter.component.ts

```diff
@@ -0,0 +1,199 @@
+  @Output() public constructedFilters = new EventEmitter<Filters>();
+
+  private previousFilters: string;
+  public model: Model = { todIgnoreDst: true };
+  public errors: {
+    todBoundaryError?: boolean;
+  } = {};
+
+  public ngAfterViewInit(): void {
+    this.form.valueChanges
+      .pipe(
+        debounceTime(defaultDebounceTime),
+        distinctUntilChanged(),
+        takeUntil(this.unsubscribe)
+      )
+      .subscribe((model: Model): void => this.updateHref(model));
+  }
+
+  public updateHref(model: Model): void {
+    // validate that the dates are a valid format & are within bounds before sending an API filter request
+    // has the value of null if there is no update to the filters
+    const newFilters = this.generateFilters(model);
+
+    // if the new filters will be the same as the old filters, do not emit a filter update event
+    if (newFilters && this.model) {
+      this.constructedFilters.emit(newFilters);
+    }
+  }
+
+  public isInvalidDate(date: string | NgbDate): boolean {
+    if (!date) {
+      return false;
+    } else if (typeof date === "string") {
+      return true;
+    }
+
+    const isValidDate = this.calendar.isValid(date)
+    return date?.year.toString().length < 4 || !isValidDate;
+  }
+
+  /**
+   * Validates that the start date comes before the end date
+   *
+   * @param formStartDate A date object of type `NgbDate` supplied by the bootstrap date input representing the start date
+   * @param formEndDate A date object of type `NgbDate` supplied by the bootstrap date input representing the end date
+   * @returns Whether the start date comes before the end date, or the two dates are equal
+   */
+  public dateIsInBoundary(formStartDate: NgbDate, formEndDate: NgbDate): boolean {
+    // if the user has not set a start date and end date, we know that the date range is not defined and therefore, cannot be out of bounds
+    if (!isInstantiated(formStartDate) || !isInstantiated(formEndDate)) {
+      return true;
+    }
+
+    // converting to a date object allows the comparison of dates using boolean operators
+    const startDate = new Date(formStartDate.year, formStartDate.month, formStartDate.day);
+    const endDate = new Date(formEndDate.year, formEndDate.month, formEndDate.day);
+
+    return startDate <= endDate;
+  }
+
+  private generateFilters(model: Model): Filters | null {
+    const filter: InnerFilter = {};
+
+    if (this.site) {
+      filter["sites.id"] = { eq: this.site.id };
+    } else if (this.region) {
+      filter["regions.id"] = { eq: this.region.id };
+    } else if (this.project) {
+      filter["projects.id"] = { eq: this.project.id };
+    }
+
+    this.setDateFilter(filter, model);
+    this.setTimeOfDayFilter(filter, model);
+
+    if (JSON.stringify(filter) === JSON.stringify({})) {
+      return null;
+    }
+
+    if (JSON.stringify({ filter }) !== this.previousFilters) {
+      this.previousFilters = JSON.stringify({filter });
+      return { filter };
+    }
```

Comment:
I chose this approach because of two reasons

1. (From my knowledge) Objects are saved as references / pointers. Therefore, to do a simple structural equality e.g. `filter === previousFilter` I would have to create and store two separate filter objects in memory (otherwise the when I update filters object, it would update previousFilters object too)

2. comparing two objects through traditional methods would compare the memory location not value e.g.

![image](https://user-images.githubusercontent.com/33742269/222034478-8c8a50e2-253b-4291-a2cd-76ed4d59bb06.png)

I could solve both these issues by instantiating two separate filter objects. But from my knowledge (may be incorrect), there is no quick way (that is less than `O(n)`) to key-value structurally compare two objects in JS

Reply:
OK; while your analysis of iteration performance might be correct it doesn't really matter when n is counted in tens (~8 keys, sometimes 20).

The other thing to keep in mind is that json.stringify does not emit keys in a particular order. 
Trivial failure example:

```js
a = { b: 2, c: 3}
b = { c: 3, b: 2}
console.log(JSON.stringify(a) == JSON.stringify(b)) // false !
```

I feel the overhead of serializing an object just to get structural equality is more than any key count consideration.

> Therefore, would have to create and store two separate filter objects in memory; 

Your current implementation is storing the previous filters as a string. That's (bullshit guess) 3 to 4 times the memory of a raw object.

> comparing two objects through traditional methods would compare the memory location not value e.g.

Yes, that's what we call reference equality: comparing references to pointers.

> . But from my knowledge (may be incorrect), there is no quick way (that is less than O(n)) to key-value structurally compare two objects in JS

N isn't really the issue. 
There are implementations like: https://www.npmjs.com/package/lodash.isequal
But I think we already use immutablejs and it feels like it should support this. https://immutable-js.com/#nested-structures

![image](https://user-images.githubusercontent.com/1452968/222044374-f3894712-bdfc-482e-8500-013f62ab7ba6.png)

But I'll tell you what, let's [measure it](https://jsbench.me/), not speculate it.

- compare ImmutableJS Maps
- to JSON.stringify (with ordered keys https://stackoverflow.com/a/64565203)

For a set of filters of various sizes.

Reply:
This is _not_ a fair test, but it is a good approximation:

![image](https://user-images.githubusercontent.com/1452968/222046166-90936c12-e493-4d22-8053-a3093ef929ec.png)

Reply:
Hmm, good points

Immutable.js's `.equals` method looks like it will be a very nice solution I skimmed over while solving this issue